### PR TITLE
Velocity SIMD CPU Runtime (Runtime + Scalar x2)

### PR DESCRIPTION
**Pull request by m4rs-mt**
This PR adds the announced SIMD-based CPU runtime fully implemented in managed code. The new `Velocity` accelerator supports most GPU kernels (except those using dynamic shared memory) and is able to utilize SIMD hardware acceleration on modern CPUs, allowing you to run your ILGPU kernels efficiently on then CPU by leveraging the implemented automatic vectorization engine.

It supports the following hardware configurations after merging *ALL* velocity PRs.

- Scalar version simulating warps of length 2 (added in this PR)
- `128bit`-based `X64 SSE` and `ARM64 Neon` instructions (also supports `M1` Macs - #769, in progress)
- `256bit`-based `X64 AVX` instructions (in progress)
- `512bit`-based `X64 AVX2` instructions (limited feature set; some functions will fallback to `256bit` registers)

Please note that this is the *initial PR* adding support for building and managing `Velocity` devices. Furthermore, it also contains the fully-featured code generator to create SIMD-based instructions out of ILGPU IR nodes. However, it does not contain any SIMD code-generation plugins for the backend code generator that will be added later on.

This PR integrates CI-support contributed by @MoFtZ in #1096.

*Note that this PR is a new version of PR #891*.

~~This PR depends on #1059, #1061, #1062, #1063, #1064, #1065, #1066, #1067, #1068, #1069, #1070, #1071, #1072, #1073, #1074, #1079, and #1081~~

Co-authored-by: MoFtZ <MoFtZ@users.noreply.github.com>

### CODE REVIEW

File: Src/ILGPU.Tests.Velocity/ILGPU.Tests.Velocity.csproj [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.7.2" />
+    <PackageReference Include="xunit" Version="2.5.0" />
```
Comment (MoFtZ) [inline Src/ILGPU.Tests.Velocity/ILGPU.Tests.Velocity.csproj:19]: ISSUE: Missing `xunit.runner.visualstudio` nuget package. This prevents the Velocity tests from being discovered in Visual Studio.

**Review by MoFtZ — CHANGES_REQUESTED** [review]
ISSUE: `VelocityUnaryFloatOperations`, `IsPredicate64` tests are failing.

image

`address.item1` = 1
`value.item1` = 1

`  System.AggregateException : One or more errors occurred. (Object reference not set to an instance of an object.)
---- System.NullReferenceException : Object reference not set to an instance of an object.
`

Tested on Intel Core i7-7700.

**Comment by m4rs-mt** [comment]
@MoFtZ That is unexpected and should clearly not happen. Let me try to replicate that.

**Comment by MoFtZ** [comment]
> @MoFtZ That is unexpected and should clearly not happen. Let me try to replicate that.

Based on #1096 and https://github.com/m4rs-mt/ILGPU/actions/runs/6168945265/job/16742300692?pr=1096, it looks like the issue is affecting Windows, but not Ubuntu?

**Comment by m4rs-mt** [comment]
I can confirm that this issue does not appear on Ubuntu x64/arm64 and on MacOS arm64.

**Comment by m4rs-mt** [comment]
@MoFtZ I addressed an issue in the `ScalarX2` code generator which know passes all tests on Windows/Intel x64.

**Review by MoFtZ — APPROVED** [review]

**Pull request merged**